common/cpp/crypto/mbedtls/sig_private_key.cpp

```diff
@@ -0,0 +1,498 @@
+ * That is, the same interface as mbedtls_ctr_drbg_random().
+ * Implemented with RandomBitString().
+ *
+ * @param unused     Unused and ignored parameter for context. Set to nullptr
+ * @param output     String containing raw binary plaintext
+ * @param output_len Length of random bit string in bytes
+ * @returns          0 on success
+ *                   MBEDTLS_ERR_CTR_DRBG_REQUEST_TOO_BIG if output_len >1024
+ *                   non-0 on other failures
+ */
+static int mbed_ctr_drbg_random_wrapper(void *unused,
+        unsigned char *output, size_t output_len) {
+    std::string s;
+
+    // Sanity check
+    if (output_len > MBEDTLS_CTR_DRBG_MAX_REQUEST)
+        return MBEDTLS_ERR_CTR_DRBG_REQUEST_TOO_BIG;
+
+    try {
+        s = ByteArrayToStr(tcf::crypto::RandomBitString(output_len));
+    } catch (const tcf::error::RuntimeError) {
+        return 1; // failure
+    }
+    memcpy(output, s.data(), output_len);
+    return 0; // success
+}   // mbed_ctr_drbg_random_wrapper()
+
+
+/**
+ * Generate ECDSA private key.
+ * Throws RuntimeError.
+ */
+void pcrypto::sig::PrivateKey::Generate() {
+    int rc;
+
+    // Allocate and initialize context
+    mbedtls_ecdsa_context *ecdsa_ctx =
+        (mbedtls_ecdsa_context *)malloc(sizeof (mbedtls_ecdsa_context));
+    if (ecdsa_ctx == nullptr) {
+        std::string msg(
+            "Crypto Error (sig::PrivateKey::Generate(): "
+            "Could not allocate memory for ECDSA key context");
+        throw Error::RuntimeError(msg);
+    }
+    mbedtls_ecdsa_init(ecdsa_ctx);
+
+    // Generate key
+    rc = mbedtls_ecdsa_genkey(ecdsa_ctx, MBEDTLS_ECP_DP_SECP256K1,
+        mbed_ctr_drbg_random_wrapper, nullptr);
+    if (rc != 0) {
+        mbedtls_ecdsa_free(ecdsa_ctx);
+        free(ecdsa_ctx);
+        std::string msg("Crypto Error (sig::PrivateKey::Generate): "
+            "mbedtls_ecdsa_genkey() failed");
+        throw Error::RuntimeError(msg);
+    }
+
+    // Free old context, if any, and set new context
+    if (private_key_ != nullptr) {
+        mbedtls_ecdsa_free((mbedtls_ecdsa_context *)private_key_);
+        free(private_key_);
+    }
+
+    private_key_ = (void *)ecdsa_ctx;
+}  // pcrypto::sig::PrivateKey::Generate
+
+
+/**
+ * Serialize ECDSA PrivateKey.
+ *
+ * For example,
+ * -----BEGIN EC PRIVATE KEY-----
+ * MIIBEwIBAQQgvIEpXzorm7Y6e0Pvzdt5hZicLG8k1+OMi0TSUbBZD0+ggaUwgaIC
+ * AQEwLAYHKoZIzj0BAQIhAP////////////////////////////////////7///wv
+ * MAYEAQAEAQcEQQR5vmZ++dy7rFWgYpXOhwsHApv82y3OKNlZ8oFbFvgXmEg62ncm
+ * o8RlXaT7/A4RCKj9F7RIpoVUGZxH0I/7ENS4AiEA/////////////////////rqu
+ * 3OavSKA7v9JejNA2QUECAQGhRANCAARxy5u39/yqw2tI98mVa4+KOnR4lAMPdFQr
+ * uTiAZ2UMH+JrTyzGoChmP7hIrxHirYc7T0hTPbN3oVgWbfQEmXsv
+ * -----END EC PRIVATE KEY-----
+ *
+ * Throws RuntimeError.
+ *
+ * @returns Serialized ECDSA private key as a string
+ */
+std::string pcrypto::sig::PrivateKey::Serialize() const {
+    static const unsigned int max_pem_len = 16000;
```

Comment:
Any specific reason to use the constant value 16000

Reply:
The longest cert I find is 4K.  So I doubled it twice for safety, which is 16K.
I moved the constant to a header file and added a comment to explain the number.